File: 213. House Robber II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];
        int prev1, cur1, prev2, cur2;
        prev1 = cur1 = prev2 = cur2 = 0;
        for (int i = 0; i < nums.size(); i++) {
            int tmp;
            if (i != nums.size() - 1) {
                tmp = cur1;
                cur1 = max(cur1, prev1 + nums[i]);
                prev1 = tmp;
            }
            if (i != 0) {
                tmp = cur2;
                cur2 = max(cur2, prev2 + nums[i]);
                prev2 = tmp;
            }
        }
        return max(cur1, cur2);
    }
};

File: java/213. House Robber II.java
// O(n)
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        return Math.max(rob(nums, 0, nums.length - 2), rob(nums, 1, nums.length - 1));
    }

    private int rob(int[] nums, int start, int end) {
        int m1 = 0, m2 = 0;
        for (int i = start; i <= end; ++i) {
            int temp = m2;
            m2 = Math.max(m2, nums[i] + m1);
            m1 = temp;
        }
        return m2;
    }
}
